[asan] Add address sanitizer support.

oleksandrivantsiv opened the pull request:
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it

#### How I did it

#### How to verify it

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


#### A picture of a cute animal (not mandatory but encouraged)

alexrallen reviewed (changes requested):
Thanks for this enhancement!

In your PR message I would prefer there be a bit more detail in the "How to Validate" section than "Run Regression"

Just briefly describe what manual testing you did and if there are any unit tests present, this will allow the reviewer to quickly understand what has already been validated prior to the review.

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+            return int(assigned_lanes_num[1:-1])
+
+        def __str__(self):
+            return "{} {} [{}] {}".format(self.port_num, self.default_speed, ", ".join(
```
I left this method only for debugging purposes. Removing it since it is not used in the code.

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+        """
+
+        # Asymmetric breakout mode
+        if re.search("\+",bmode) is not None:
```
Agree

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+    def get_config(self):
+        # Ensure that we have corret number of configured lanes
+        lanes_used = 0
+        for entrie in self._breakout_mode_entry:
```
typo, changed

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+            for port in range(entrie.port_num):
+                interface_name = PORT_STR + str(self._interface_base_id + lane_id)
+
+                if not entrie.default_speed.endswith('G'):
```
It is not part of dynamic port breakout HDL. I suggest leaving it as is

dgsudharsan replied:
Please add the logic back as discussed offline.

oleksandrivantsiv replied:
done

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+class BreakoutCfg(object):
+
+    class BreakoutModeEntry:
+        def __init__(self, port_num, default_speed, supported_speed, assigned_lanes_num=None):
```
Agree

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,162 @@ def parse_port_config_file(port_config_file):
+            if not assigned_lanes_num.startswith('(') and not assigned_lanes_num.endswith(')'):
+                raise RuntimeError("Unsupported port breakout format!")
+
+            return int(assigned_lanes_num[1:-1])
```
I don't think that this is needed. The exception that "int" raises indicates the problem.

oleksandrivantsiv commented:
@alexrallen Please review

dgsudharsan commented on src/sonic-config-engine/tests/sample_output/platform_output.json:
```diff
@@ -1,823 +1,883 @@
 {
-  "Ethernet8": {
```
Why are we having such a huge diff in this file? Shouldn't it be only for the entries you added?

oleksandrivantsiv replied:
This file has autogenerated output from sonic-cfggen utility. It was easier for me to run sonic-cfggen utility again and save the output to the file rather than edit it manually. However, it makes diff unreadable so in the last commit, I added new entries manually.

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,140 @@ def parse_port_config_file(port_config_file):
+        def __init__(self, num_ports, default_speed, supported_speed, num_assigned_lanes=None):
+            self.num_ports = int(num_ports)
+            self.default_speed = default_speed
+            self.supported_speed = set((default_speed, ))
```
To be able to compare breakout mode supported speeds I'm using "set" collection. As the input parameter set expects to have an iterable value. "default_speed" has string type. The string is iterable. If we pass "default_speed" as a parameter to set each character will be used as separated entry in the set:
```
In [2]: set("10G")
Out[2]: {'0', '1', 'G'}
```
To prevent this I'm passing a tuple with one element as an input parameter:
```
In [3]: set(("10G", ))
Out[3]: {'10G'}
```

oleksandrivantsiv commented on src/sonic-config-engine/portconfig.py:
```diff
@@ -172,85 +172,140 @@ def parse_port_config_file(port_config_file):
+                return False
 
-            ports[intf_name]['index'] = index.split(",")[alias_position]
+    def __init__(self, name, bmode, capabilities):
```
done

alexrallen reviewed (approved):
All changes look good, thanks.

oleksandrivantsiv committed:
[asan] Add address sanitizer support.

Implement infrastructure that allows enabling address sanitizer
for docker containers. Enable address sanitizer for SWSS container.

liat-grozovik commented on files/build_templates/docker_image_ctl.j2:
```diff
@@ -302,6 +302,10 @@ start() {
     fi
+
+    {%- if enable_asan == "y" %}
+    mkdir -p /var/log/asan
```
is there way to have this as a constant/define to be shared with swss on another PR you created?

oleksandrivantsiv replied:
I changed the way how to pass address sanitizer options to applications. Now they come via the environment variable defined in this repo (in supervisord.conf file).

liat-grozovik commented on files/build_templates/docker_image_ctl.j2:
```diff
@@ -486,6 +493,8 @@ stop() {
     {%- elif docker_container_name == "teamd" %}
     # Longer timeout of 60 sec to wait for Portchannels to be cleaned.
     /usr/local/bin/container stop -t 60 $DOCKERNAME
+    {%- elif docker_container_name == "swss" and  enable_asan == "y" %}
```
extra space after enable_asan

oleksandrivantsiv replied:
nice catch, thanks!

liat-grozovik commented on files/build_templates/docker_image_ctl.j2:
```diff
@@ -486,6 +493,8 @@ stop() {
     # Longer timeout of 60 sec to wait for Portchannels to be cleaned.
     /usr/local/bin/container stop -t 60 $DOCKERNAME
+    {%- elif docker_container_name == "swss" and  enable_asan == "y" %}
+    /usr/local/bin/container stop -t 60 $DOCKERNAME
```
can you explain the need for this extra time?

oleksandrivantsiv replied:
It is needed to give enough time to address sanitizer to generate the report. Report generation is a time-consuming operation that varies based on the amount of used memory.

liat-grozovik commented on rules/config:
```diff
@@ -208,3 +208,6 @@ BUILD_MULTIASIC_KVM = n
 INCLUDE_MUX = y
+
+# ENABLE_ASAN - enable address sanitizer
+ENABLE_ASAN ?= n
```
should it be '= n' only?

oleksandrivantsiv replied:
'?=' allows overriding value via an environment variable. Eg. `ENABLE_ASAN=y make ...`

liat-grozovik commented on slave.mk:
```diff
@@ -184,6 +184,13 @@ ifeq ($(SONIC_INCLUDE_MUX),y)
 endif
 
+ifeq ($(ENABLE_ASAN),y)
+	ifneq ($(CONFIGURED_ARCH),amd64)
```
is it known it is not supported on this arch? or it is only because we cannot validate it?

oleksandrivantsiv replied:
it is known that arm64 architecture in Linux is not supported, the matrix with supported architectures is here:
https://github.com/google/sanitizers/wiki/AddressSanitizer

dgsudharsan commented:
@oleksandrivantsiv Have you checked whether techsupport includes the ASAN logs? I believe by default all logs under (/var/log) will be saved but please confirm.
https://github.com/Azure/sonic-utilities/blob/d9f3afe5b34ef0f03f4137f607bc73bc625631ed/scripts/generate_dump#L959

dgsudharsan commented:
@oleksandrivantsiv Having ASAN enabled would likely increase fastboot time and also have impacts on downtime during config reload and other scenarios correct? Should we document it?

oleksandrivantsiv commented:
> @oleksandrivantsiv Have you checked whether techsupport includes the ASAN logs? I believe by default all logs under (/var/log) will be saved but please confirm. https://github.com/Azure/sonic-utilities/blob/d9f3afe5b34ef0f03f4137f607bc73bc625631ed/scripts/generate_dump#L959

Yes, it includes ASAN logs

oleksandrivantsiv commented:
> @oleksandrivantsiv Having ASAN enabled would likely increase fastboot time and also have impacts on downtime during config reload and other scenarios correct? Should we document it?

in theory, it should not affect fastboot time. When fastboot is initiated swss service is killed and reports are not generated: 
```
    # For WARM/FAST boot do not perform service stop
    if [[ x"$WARM_BOOT" != x"true" ]] && [[ x"$FAST_BOOT" != x"true" ]]; then
        /usr/bin/${SERVICE}.sh stop $DEV
        debug "Stopped ${SERVICE}$DEV service..."
    else
        debug "Killing Docker swss..."
        /usr/bin/docker kill swss &> /dev/null || debug "Docker swss is not running ($?) ..."
    fi
```
I'll run tests and collect results to understand what we have in practice. If time is affected I'll update documentation
